--- pexec.pl.patch4	2011-09-09 16:18:44.000000000 -0600
+++ pexec.pl	2011-09-09 16:37:16.457018640 -0600
@@ -366,6 +366,23 @@
    pod2usage("\n$prog: no machine group specified\n");
 }
 
+if (-e '/usr/bin/wwlist') {
+   local $ww_list = '/usr/bin/wwlist';
+   local $ww_down = '-d';
+   local $ww_idle = '-i';
+   local $ww_ready = '-r';
+} elsif (-e '/usr/bin/wwnodes') {
+   local $ww_list = '/usr/bin/wwnodes';
+   local $ww_down = '-d';
+   local $ww_idle = '-i';
+   local $ww_ready = '';
+} else {
+   local $ww_list = '';
+   local $ww_down = '';
+   local $ww_idle = '';
+   local $ww_ready = '';
+}
+
 local %NETGROUP = ();
 &CacheNetgroups(\%NETGROUP);
 local @MachineList = &GetMachines(\%NETGROUP,\@opt_m);
@@ -885,6 +902,42 @@
 		  }
 	       }
 	    }
+	 } elsif ($name =~ /^allup$/i) {
+	    -x $ww_list || (push(@Machines,$name) , next M);
+	    open(WWLIST, "$ww_list $ww_ready |") ||
+	       (push(@Machines,$name) , next M);
+	    my $header = <WWLIST>;
+	    foreach my $node (<WWLIST>) {
+	       chop $node;
+	       if ($node =~ /^(\S+).*$/) {
+		  push(@Machines, $1);
+	       }
+	    }
+	    close WWLIST;
+	 } elsif ($name =~ /^alldown$/i) {
+	    -x $ww_list || (push(@Machines,$name) , next M);
+	    open(WWLIST, "$ww_list $ww_down |") ||
+	       (push(@Machines,$name) , next M);
+	    my $header = <WWLIST>;
+	    foreach my $node (<WWLIST>) {
+	       chop $node;
+	       if ($node =~ /^(\S+).*$/) {
+		  push(@Machines, $1);
+	       }
+	    }
+	    close WWLIST;
+	 } elsif ($name =~ /^allidle$/i) {
+	    -x $ww_list || (push(@Machines,$name) , next M);
+	    open(WWLIST, "$ww_list $ww_idle |") ||
+	       (push(@Machines,$name) , next M);
+	    my $header = <WWLIST>;
+	    foreach my $node (<WWLIST>) {
+	       chop $node;
+	       if ($node =~ /^(\S+).*$/) {
+		  push(@Machines, $1);
+	       }
+	    }
+	    close WWLIST;
 	 } else {					# a single host
 	    push(@Machines,$name) unless ($name =~ /^$/);
 	 }
@@ -1076,6 +1129,9 @@
 
 Include I<host> in the machine list.  Valid names are: hosts, netgroups,
 ranges, or a path to files containing these entities (newline separated).
+You may also use special keywords "I<allup>", "I<alldown>" and "I<allidle>" in
+the machine specification and I<pexec> will contact the local I<warewulfd>
+to populate the list with node names posessing such a characteristic.
 Each I<host> specification may be inter-mixed however always comma
 separated.  Netgroup names are invoked by using the "@" symbol, which acts
 as an intersection operator.  A single netgroup can be designated with a
