--- pexec.pl.patch0	2014-01-10 14:14:35.224454664 -0700
+++ pexec.pl	2014-01-10 14:22:02.661461383 -0700
@@ -35,6 +35,7 @@
 $ThisMachine =~ s/\..*//;
 
 our $prompt = $ENV{'prompt'} || '# ';
+our $xcatroot = $ENV{'XCATROOT'} || '/opt/xcat';
 
 unshift (@ARGV, split(/\s+/, $ENV{'PEXEC'})) if (exists $ENV{'PEXEC'});
 
@@ -694,20 +695,20 @@
 				# look-ahead assertion)
 
    foreach my $machines (@{ $aref }) {
-      $machines =~ s/\s+//go;				# kill whitespace
-
-      foreach my $name (split(/$comma/,$machines)) {
-#                        prefix  [NR , digit] suffix
-	 if ($name =~ /^([^\[]+)\[([\d,-]+)\]([^\[]*)$/) {
-	    my $match  = $&;				# translate
-	    my $prefix = $1;				#   p[1-10,13]s
-	    my $suffix = $3;				# to
-	   (my $string = $2) =~				#   p1s-p10s,p13s
-	      s/(\d+)/$prefix$1$suffix/g;		# for prefix 'p' and
-	    $machines =~ s/\Q$match\E/$string/;		# suffix 's'
+      if (open(NR, "$xcatroot/bin/nodels $machines 2>/dev/null |")) {
+	 my @Range = ();
+	 foreach my $node (<NR>) {
+	    chomp $node;
+	    push(@Range, $node);
+	 }
+	 close NR;
+	 if (scalar @Range) {
+	    push(@Machines, @Range);
+	    next;
 	 }
       }
 
+      $machines =~ s/\s+//go;				# kill whitespace
 M:    foreach my $name (split(/$comma/,$machines)) {
 	 my @Dash = split('-', $name);
 	 if ($name =~ /^\@([^\@]+)$/) {			# a lone netgroup
@@ -1072,11 +1073,7 @@
 
 Include I<host> in the machine list.  Valid names are: hosts, netgroups,
 ranges, or a path to files containing these entities (newline separated).
-You may also use special keywords "I<allrespond>" (a.k.a "I<allup>"),
-"I<allnorespond>" (a.k.a. "I<allnotup>"), "I<alldown>", "I<allidle>",
-"I<allalloc>", "I<alldrain>", "I<alldrained>" and "I<allcomp>" in the machine
-specification and B<pexec> will contact the local I<slurmd> to populate the
-list with node names posessing such a characteristic.
+You may also use any valid xCAT node range as an argument (see noderange(3)).
 Each I<host> specification may be inter-mixed however always comma
 separated.  Netgroup names are invoked by using the "@" symbol, which acts
 as an intersection operator.  A single netgroup can be designated with a
@@ -1328,6 +1325,11 @@
 Take B<pexec> options from this environment variable, which is parsed
 before the command line (and thereby overridden).
 
+=item B<XCATROOT>
+
+Use the directory specified by this environment variable as the xCAT root
+from which a noderange is calculated (default: /opt/xcat).
+
 =back
 
 =head1 FILES
@@ -1372,7 +1374,7 @@
 
 =head1 SEE ALSO
 
-gethostbyname(3), netgroup(5), rsh(1), rcp(1),
+gethostbyname(3), netgroup(5), noderange(3), rsh(1), rcp(1),
 rsync(1), ssh(1), scp(1)
 
 =head1 AUTHOR
